clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -278,7 +282,7 @@ public synchronized boolean isAborting() {
     }
 
     synchronized boolean isInTransaction() {
-        return currentState == State.IN_TRANSACTION || isCompletingTransaction();
+        return currentState == State.IN_TRANSACTION || isCompletingTransaction() || hasAbortableError();
```

Comment:
This bug is a bit insidious. We use `isInTransaction` when dequeuing records from the `RecordAccumulator` to determine if the batch should be transactional. So if we continued trying to send after an abortable error, we could lose the transactional flag.